File: solutions/kotlin/run-length-encoding/1/src/main/kotlin/RunLengthEncoding.kt
object RunLengthEncoding {

    fun encode(input: String): String {
        if (input.isEmpty()) return ""
        
        val result = StringBuilder()
        var count = 1
        
        for (i in 1 until input.length) {
            if (input[i] == input[i - 1]) {
                count++
            } else {
                if (count > 1) result.append(count)
                result.append(input[i - 1])
                count = 1
            }
        }
        
        if (count > 1) result.append(count)
        result.append(input.last())
        
        return result.toString()
    }

    fun decode(input: String): String {
        val result = StringBuilder()
        var count = 0
        
        for (char in input) {
            if (char.isDigit()) {
                count = count * 10 + (char - '0')
            } else {
                val repeatCount = if (count == 0) 1 else count
                result.append(char.toString().repeat(repeatCount))
                count = 0
            }
        }
        
        return result.toString()
    }
}

File: solutions/julia/run-length-encoding/1/run-length-encoding.jl
function encode(s)
    
    matches = eachmatch(r"(.)\1*", s)
    
    result = IOBuffer()
    for m in matches
        str = m.match
        len = length(str)
    
        if len > 1
            print(result, len)
        end
        print(result, str[1])
    end
    
    return take!(result) |> String
end

function decode(s)

    matches = eachmatch(r"(\d*)(.)", s)
    
    result = IOBuffer()
    for m in matches
        count_str = m.captures[1]
        char = m.captures[2]
        
   
        count = isempty(count_str) ? 1 : parse(Int, count_str)
        print(result, char ^ count)
    end
    
    return take!(result) |> String
end
